### PR TITLE
feat(syrup-frame): syrup-frame package and opt-in framing for OCapN TCP-for-testing

### PR DESCRIPTION
**Pull request by kriscendobot**
Refs: #29 #108

## Description

Adds a new workspace package `@endo/syrup-frame`: a sibling of `@endo/netstring` whose framing omits the trailing `,` separator, so each framed payload on the wire is literally a Syrup byte-string record (`<length>:<payload>`). The exported surface is symmetric with `@endo/netstring`: `makeSyrupReader` and `makeSyrupWriter`, including the `chunked` zero-copy writer mode.

The OCapN TCP-for-testing netlayer gains an opt-in `framing` option. The default `'none'` preserves the current wire format used by the `ocapn/ocapn-test-suite` Python `testing_only_tcp` netlayer (a single syrup-encoded record per write, no length prefix). Passing `framing: 'syrup'` wraps each message in the new framing so the transport becomes robust to TCP chunk boundaries that split a single OCapN message, and two peers that both opt in share a single length-prefixed primitive with the syrup payload format itself.

The design rationale (why a separate package rather than an option on `@endo/netstring`, the OCapN TCP-for-testing scope, alternatives considered) lives in the sibling design PR #108.

### Security Considerations

The package parses untrusted byte streams. The reader rejects malformed prefixes, oversize length declarations (configurable via `maxMessageLength`, default 999_999_999), missing colon, empty prefix, and trailing characters after a complete frame. No new authority is introduced beyond the `Uint8Array` reader/writer pair the package exports. All exported factories, the inner generator, and the returned writer object are `harden()`-ed. The opt-in TCP netlayer framing does not alter the authority boundary of the netlayer; it changes only how bytes are grouped on the wire.

### Scaling Considerations

`maxMessageLength` bounds reader memory at the message granularity. The chunked writer mode avoids a prefix-plus-payload buffer copy. The netlayer framing wrapper adds one length-prefix per message and no other steady-state cost.

### Documentation Considerations

The new package's `README.md` describes the grammar, the asymmetry against `@endo/netstring`, the OCapN TCP-for-testing scope, the API, and the options. Initial-release notes ship in the package changeset. The TCP-netlayer framing option ships with its own changeset describing the opt-in. No deployed-data migration is required.

### Testing Considerations

The new package carries reader, writer, chunked-write, concurrent-write, byte-by-byte fragmentation, trailing-comma rejection, and `writer.throw` forwarding tests, run under both `lockdown` and `noop-harden` configurations. An additional netlayer test exercises a syrup-framed TCP handshake end to end against the JS netlayer.

### Compatibility Considerations

New package; no backwards-compatibility concerns. `@endo/netstring` is unmodified; existing callers are unaffected. The Endo daemon's `tcp+netstring+json+captp0` transport remains netstring-compliant. The TCP-for-testing netlayer's default framing is unchanged; the new behavior is opt-in.

### Upgrade Considerations

No upgrade action required. Initial release of a new package and an additive, opt-in option on an existing test-only netlayer.

### CODE REVIEW

**Review by kriscendobot — COMMENTED** [review]
## Panel review: 12 perspectives

The panel reviewed the `@endo/syrups` package as introduced in this
PR. Findings are organized into must-fix / should-fix / out-of-scope
at the bottom; the per-perspective blocks below are the working
notes.

### Reviewer 1: Correctness

**Verdict:** comment-only

**Findings:**

- `reader.js` line 101 uses `buffer.length >= remainingDataLength`
  (the `>=` is correct for syrups because the payload has no
  trailing separator byte; netstring's `>` reserves a byte for the
  comma). This asymmetry is the only behavioral departure and is
  load-bearing; a comment cross-references the design.
- `reader.js` removed the `Number.isNaN` guard that
  `@endo/netstring/reader.js` keeps (lines 77-80 of the netstring
  reader). The comment on lines 81-84 of the syrups reader argues
  the prefix is always a valid finite non-negative integer because
  the inner loop only pushes ASCII digits. This is true, but the
  guard was load-bearing in netstring as defense-in-depth against
  a future refactor of the inner loop. Removal is defensible
  (cleaner did this in `f468e88821`); keep an eye on it if the
  inner loop ever changes.
- `writer.js` line 14 computes `${length | 0}:` with a bitwise OR
  that truncates `length` to a 32-bit signed integer. `1 << 31`
  becomes `-2147483648`, and a `messageLength` greater than
  `2 ** 31 - 1` underflows to a negative-prefix string. This
  matches `@endo/netstring`'s identical behavior, but the panel
  flags it as inherited risk worth documenting.

**Notes (out of scope):**

- The same per-chunk-promise rejection-handler fix could be
  back-ported to `@endo/netstring/writer.js`, which currently
  lacks it. Out of scope for this PR but worth a follow-up.

### Reviewer 2: Test coverage

**Verdict:** comment-only

**Findings:**

- 27 tests cover the reader (well-formed input, fragmentation,
  error paths), the writer (chunked / unchunked, concurrent
  writes, close-anywhere), `writer.throw`, and the iterable
  contract.
- The "trailing comma after frame is rejected (not silently
  consumed)" test is the load-bearing assertion that distinguishes
  syrups from netstring. The test actually exercises the
  `Invalid syrups length prefix` path because `,` is not a digit;
  the test would pass even if the reader naively allowed any
  non-digit transition. A second variant feeding `5:hello5:world`
  would confirm that two adjacent valid frames *are* read
  correctly (which the existing `read messages divided over chunk
  boundaries` test does cover, indirectly).
- No test exercises the `maxMessageLength` boundary condition at
  exactly the limit; both `fails reading too long prefix` and
  `fails reading if message length over max` exercise overshoots
  but not the boundary itself.
- No test for `return()` on the writer outside the round-trip
  contexts. Coverage is OK because `writer.return()` is a
  pass-through.

**Notes (out of scope):**

- A property-based test (e.g., `@endo/random/fast-check`) over
  random byte sequences would be valuable but is its own follow-up.

### Reviewer 3: TypeScript / type-system enforcement

**Verdict:** comment-only

**Findings:**

- `reader.js` and `writer.js` both start with `// @ts-check` and
  carry `@param`, `@returns`, and `@type` annotations.
- `typeCoverage.atLeast: 100` in `package.json` matches netstring;
  `yarn lint:types` passes locally.
- The test file uses `// @ts-nocheck`. This matches netstring's
  test file conventions (which also uses `@ts-nocheck`), so the
  convention is uniform across the package family. The trade-off
  (avoid having to type the test fixtures fully) is reasonable.
- The writer's `messageChunks` parameter is reassigned inside
  `next()` (`if (!Array.isArray(messageChunks)) messageChunks =
  [messageChunks]`). Reassignment of a parameter narrowed by
  `Array.isArray` is fine here. No findings.

### Reviewer 4: API stability / breaking-change accuracy

**Verdict:** approve

**Findings:**

- New package, no prior API to break. Changeset declares `patch`
  initial release at `0.1.0`.
- Exported surface is symmetric with `@endo/netstring`:
  `makeSyrupsReader`, `makeSyrupsWriter`. No legacy aliases.
- Subpath exports under `./reader.js` and `./writer.js` match
  netstring's convention.

### Reviewer 5: Diff hygiene

**Verdict:** approve

**Findings:**

- Two commits: the `feat(syrups)` substantive commit and the
  separate `chore: Update yarn.lock`. Lock file in its own commit
  per the repo convention.
- All paths under `packages/syrups/` plus `.changeset/syrups-initial.md`
  plus `yarn.lock`. No unrelated files.
- No agent-attribution trailers.

### Reviewer 6: Error-message quality

**Verdict:** comment-only

**Findings:**

- All four reader error messages include the `name` and `offset`,
  which is the established netstring pattern.
- `Invalid syrups length prefix "1.0"` etc. quote the offending
  bytes via `JSON.stringify(String.fromCharCode(...))`. Readable.
- `Syrups message too big (length 1000) at offset 0 of <unknown>`
  reports the size. Good.
- One minor: the `Invalid` errors use `Error(...)` without a
  capability-style `makeError` from `@endo/errors`. This is
  consistent with `@endo/netstring`, but if the panel is
  prescribing a forward style the new package is the place to
  introduce it. The panel does not insist; the consistency with
  netstring is the stronger argument for now.

### Reviewer 7: Performance

**Verdict:** approve

**Findings:**

- The reader's zero-copy fast path (`buffer.length >=
  remainingDataLength` and no pre-existing `dataBuffer`) yields
  the input chunk's subarray directly without an allocation.
- The writer's `chunked` mode avoids the prefix+payload buffer
  copy; the unchunked mode allocates one combined buffer.
- The unchunked writer path uses `Uint8Array.set` for both prefix
  and chunks; this is the standard approach.

### Reviewer 8: Naming and prose

**Verdict:** approve

**Findings:**

- Identifiers consistently use `Syrups` (plural-of-format), matching
  the design's recommendation and the queued `@endo/cbors` naming.
- Error messages use lowercase `syrups` ("Too long syrups length
  prefix"), which is the consistent prose form.
- README.md sentence-per-line, no em-dashes, prose under 80-100
  cols. Consistent with repo style.

### Reviewer 9: Changeset accuracy and severity

**Verdict:** approve

**Findings:**

- `.changeset/syrups-initial.md` declares `'@endo/syrups': patch`
  for an initial release. `patch` rather than `minor` is the
  established convention for an initial-publication changeset
  in this monorepo.
- The changeset prose lists the new package, the exported
  identifiers, the `chunked` mode, and the "not yet wired into
  any OCapN netlayer" caveat. Accurate and complete.

### Reviewer 10: Backwards compatibility

**Verdict:** approve

**Findings:**

- New package, no existing users. No backwards-compatibility
  concerns.
- Existing `@endo/netstring` is unmodified; callers of netstring
  are unaffected.

### Reviewer 11: Documentation / metadata

**Verdict:** comment-only

**Findings:**

- `README.md` covers grammar, asymmetry vs netstring, scope, API,
  options, and links to the design.
- `CHANGELOG.md` carries the initial-release entry.
- `package.json` metadata is complete: keywords, homepage,
  repository directory, exports, files, sesAvaConfigs, typeCoverage.
- `LICENSE` and `SECURITY.md` are present and match the repo
  convention.
- One observation: the design at
  `designs/ocapn-tcp-syrups-framing.md` is referenced from
  `README.md` and from the source-file headers. That file ships
  in the sibling design PR (#108), not in this PR. A reader of
  this PR's history at master will not find the design file by
  following the link until #108 lands. Acceptable for the split,
  but worth noting in the PR body (which the PR body already does:
  "design half of #29 ships separately at #108").

### Reviewer 12: Security / capability surface

**Verdict:** comment-only

**Findings:**

- Reader bounds memory by `maxMessageLength` (default 999999999, ~1
  GB). The default is large; a deployment exposing this reader to
  untrusted input should override it.
- Reader rejects malformed prefixes, missing colon, empty prefix,
  and oversize length declarations.
- All exported factories, the inner generator, and the returned
  writer object are `harden()`-ed. No mutation surface.
- Writer's `throw()` forwards to the underlying output's `throw()`.
  No additional authority is conferred; the writer is a transparent
  framing layer.
- The reader's accept loop processes bytes from an untrusted
  source. The single-byte fragmentation test exercises arbitrary
  attacker-chosen split points without surfacing any allocation
  unboundedness; the `dataBuffer` is allocated once per message at
  exactly the declared length.

## Aggregated must-fix list

None. The 27 tests pass under both `lockdown` and `noop-harden`
configurations; lint, type check, and format are clean.

## Should-fix in this PR

None raised by two or more reviewers.

## Out of scope / follow-up

- Per-chunk-promise rejection handler back-ported to
  `@endo/netstring/writer.js` (Reviewer 1, Reviewer 12 indirectly).
  The fix landed only on the syrups fork; the upstream peer
  retains the original behavior.
- Property-based reader test using `@endo/random/fast-check` over
  random byte sequences (Reviewer 2).
- Test for `maxMessageLength` exactly at the boundary (Reviewer 2).
- Two-adjacent-frames variant of the trailing-comma-rejection test
  to confirm the boundary detection isolates the malformed prefix
  rather than masking adjacent valid frames (Reviewer 2).
- Default `maxMessageLength` of ~1 GB is large; a follow-up could
  consider tightening the default for production deployments
  exposing the reader to untrusted input (Reviewer 12).
- Adopt `makeError` / `q` from `@endo/errors` for the reader's
  thrown errors, in lockstep with backporting the same change to
  `@endo/netstring` for parity (Reviewer 6).

File: packages/syrups/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+`@endo/syrups` is a deliberate departure, named for its purpose.
+
+## Scope
+
+This package is an Endo-internal variant intended for use under a
+distinct OCapN network identifier (e.g., `tcp+syrups`).
+The existing `@endo/netstring` package remains the canonical netstring
+implementation and continues to serve the Endo daemon's
+`tcp+netstring+json+captp0` transport and any other caller that wants
+strict netstring compliance.
```
Comment (kriskowal) [inline packages/syrups/README.md]: This is actually for testing interoperability with the OCapN TCP-for-testing protocol, which has moved to adopt this framing protocol. It is among the protocols in consideration for the OCapN pre-standards group  at time of writing.

**Review by kriskowal — CHANGES_REQUESTED** [review]
Small change requested for scope documentation. And, please rewrite the PR description to capture the substance and not the process.

File: packages/syrups/README.md (hunk 1 above)
Comment (kriscendobot) [reply packages/syrups/README.md]: Addressed in d50c7b7673: reframed the README scope section, the CHANGELOG entry, and the changeset prose to describe the package as the framing the OCapN TCP-for-testing protocol has adopted (and one of the protocols under consideration in the OCapN pre-standards group), rather than as a general-purpose comma-less netstring or a future `tcp+syrups` netlayer.

File: designs/hardened-text-codecs-shim.md
Comment (kriskowal) [inline designs/hardened-text-codecs-shim.md:1]: Stray, unrelated designs. Please remove both.

**Review by kriskowal — COMMENTED** [review]
Please dispatch a subagent to verify by comparing the respective modules line-by-line to the implementation of netstring, that these have not diverged more than necessary to accommodate the loss of a coma on the wire.

**Review by kriskowal — CHANGES_REQUESTED** [review]
Please add to this PR an integration with OCapN, such that this is a message framing that can be integrated in the TCP for testing netlayer. The Python test suite may or may not have migrated yet. Please check. If we can jump directly to using Syrups, please do. If not, please provide a flag. Please also dispatch a subagent to fork the Python test suite and update it to use Syrups message framing, if it does not already.

**Comment by kriskowal** [comment]
I’ve granted @kriscendobot write access to an endojs fork https://github.com/endojs/ocapn-test-suite. Be sure to use the hash pinned from here as a baseline. Do not open a PR upstream.

**Comment by kriskowal** [comment]
Ferried upstream as endojs/endo#3256 (https://github.com/endojs/endo/pull/3256).

**Comment by kriskowal** [comment]
PR title and description need an update.

**Comment by kriskowal** [comment]
Re-ferried upstream to incorporate the package rename (`@endo/syrups` -> `@endo/syrup-frame`) and the thousands-separator chore. New upstream head: `acddddba16d524c8eb16d71fb35c43f34fa491ae`.

**Comment by kriskowal** [comment]
The description is stale here, still mentioning syrups.

**Comment by kriskowal** [comment]
Please search for lingering remnants of the term “syrups” and replace them with “syrup-frame”.

**Comment by kriskowal** [comment]
And take care to reduce syrups-frame or syrup-frame-frame to syrup-frame again, as might occur from find-and-replace drift.

File: packages/ocapn/src/netlayers/tcp-test-only.js [hunk 2]
```diff
@@ -3,14 +3,35 @@
...
+ *   `testing_only_tcp` netlayer, which writes a syrup-encoded
+ *   record with `sendall` and reads one back with
+ *   `syrup.syrup_read` (no length prefix on the wire).
+ * - `'syrups'`: each message is wrapped in the
+ *   `<length>:<payload>` framing implemented by `@endo/syrup-frame`.
+ *   Adds robustness against TCP chunk boundaries that split a
+ *   single OCapN message and works with peers that opt in to the
+ *   same framing.
+ *
+ * @typedef {'none' | 'syrups'} TcpTestOnlyFraming
```
Comment (kriskowal) [inline packages/ocapn/src/netlayers/tcp-test-only.js]: Call it syrup, please. Not syrups.

File: .changeset/ocapn-tcp-syrups-framing.md
Comment (kriskowal) [inline .changeset/ocapn-tcp-syrups-framing.md:1]: Please s/syrups/syrup in the file name. Please address syrups in the description: should be simply framing: syrup.

**Review by kriskowal — CHANGES_REQUESTED** [review]

**Comment by kriskowal** [comment]
Please fix and shepherd.

**Comment by kriskowal** [comment]
Please fetch and rebase on actual/llm. Then, retcon.

**Comment by kriskowal** [comment]
Please rebase on actual/master for the guile CI job fix. Then retcon the commits.

**Comment by kriskowal** [comment]
Re-ferried over endojs/endo#3256. New upstream head: `b5c316865`.

Restructured the prior single squash into two commits to mirror the bot-side restructure:

1. `feat(syrup-frame): add @endo/syrup-frame package`
2. `feat(ocapn): add opt-in syrups framing to TCP-testing netlayer`

**Comment by kriskowal** [comment]
Fast-forward appended to endojs/endo#3256: upstream head is now b65072faf. Single yarn.lock commit cfa440f2 ferried with author rewrite; the prior two commits already on the upstream tip were not re-applied. No force-push.

**Comment by kriskowal** [comment]
We need to sync actual/master to master, then rebase this change on master. Then, we need to retcon the PR.

File: packages/ocapn/src/netlayers/tcp-test-only.js (hunk 2 above)
Comment (kriscendobot) [reply packages/ocapn/src/netlayers/tcp-test-only.js]: Addressed in 5d47bd0f6 (chore(syrup-frame): rename framing value, identifiers, and changesets to syrup (singular)). The typedef is now `'none' | 'syrup'`, the runtime guard checks against `'syrup'`, and `makeSyrupsReader`/`makeSyrupsWriter`/`makeSyrupsDeframer`/`makeSyrupsWritingSocketOperations` are now `makeSyrupReader`/`makeSyrupWriter`/`makeSyrupDeframer`/`makeSyrupWritingSocketOperations`.

File: .changeset/ocapn-tcp-syrups-framing.md
Comment (kriscendobot) [reply .changeset/ocapn-tcp-syrups-framing.md:1]: Addressed in 5d47bd0f6 (chore(syrup-frame): rename framing value, identifiers, and changesets to syrup (singular)). Renamed `.changeset/ocapn-tcp-syrups-framing.md` to `.changeset/ocapn-tcp-syrup-framing.md` and reworded the body so the option reads `framing: 'syrup'`. Also renamed `.changeset/syrups-initial.md` to `.changeset/syrup-initial.md` for the same reason.

File: packages/ocapn/src/netlayers/tcp-test-only.js [hunk 3]
```diff
@@ -46,6 +67,104 @@ const makeSocketOperations = (socket, writeLatencyMs) => {
...
+ * Wraps `socketOps` so that `write(bytes)` emits a syrup-framed
+ * record (`<length>:<payload>`) instead of raw bytes. Uses the
+ * `@endo/syrup-frame` writer over a tiny synchronous sink so framing
+ * stays inline with the underlying socket write.
+ *
+ * @param {SocketOperations} socketOps
+ * @returns {SocketOperations}
+ */
+const makeSyrupWritingSocketOperations = socketOps => {
+  /** @type {import('@endo/stream').Writer<Uint8Array, undefined>} */
```
Comment (kriskowal) [inline packages/ocapn/src/netlayers/tcp-test-only.js]: Note that we prefer `@import` and evidently don’t say so loudly enough in the style guide. Please also add this to the style juror’s checklist or carve a new juror role. This role must be used by builders. Consider creating a skill with a script for efficiently discovering violations, or other automation we can propose for CI.

File: packages/ocapn/src/netlayers/tcp-test-only.js [hunk 4]
```diff
@@ -46,6 +67,104 @@ const makeSocketOperations = (socket, writeLatencyMs) => {
...
+      for await (const frame of syrupReader) {
+        if (closed) {
+          break;
+        }
+        // The syrup reader may yield a `subarray()` of an internal
+        // buffer; downstream OCapN syrup decoding requires a
+        // zero-`byteOffset` view, so allocate a fresh copy before
+        // dispatching.
+        const owned = new Uint8Array(frame.length);
+        owned.set(frame);
```
Comment (kriskowal) [inline packages/ocapn/src/netlayers/tcp-test-only.js]: Consider `slice()`

File: packages/ocapn/src/netlayers/tcp-test-only.js [hunk 5]
```diff
@@ -163,6 +311,22 @@ export const makeTcpNetLayer = async ({
...
+   * length-prefixed frame on the wire.
+   * @param {net.Socket} socket
+   * @returns {SocketOperations}
+   */
+  const makeFramedSocketOperations = socket => {
+    const rawOps = makeSocketOperations(socket, writeLatencyMs);
+    if (framing === 'syrup') {
+      return makeSyrupWritingSocketOperations(rawOps);
+    }
+    return rawOps;
```
Comment (kriskowal) [inline packages/ocapn/src/netlayers/tcp-test-only.js:325]: We should leave a note that this is debt that needs to be cleaned up. The default socket operations should use `syrup` and we only allow `none` for testing with the existing Python suite, which is known to be inadequate against the possibility of a TCP chunk getting split across packets.

File: packages/ocapn/test/netlayer-tcp-syrup.test.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+const COMMA = ','.charCodeAt(0);
+const COLON = ':'.charCodeAt(0);
+const ZERO = '0'.charCodeAt(0);
+const NINE = '9'.charCodeAt(0);
+
+/**
+ * Establishes a TCP server that captures the first chunk of bytes
+ * received from any connection, then closes the connection. Used to
+ * peek at the wire format of an outgoing handshake from the JS
+ * netlayer.
```
Comment (kriskowal) [inline packages/ocapn/test/netlayer-tcp-syrup.test.js]: “the JS netlayer” is an odd proper noun. All the netlayers will have JS implementations, presumably, so there is a distinction without a difference. Please reword.

File: packages/ocapn/test/netlayer-tcp-syrup.test.js [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  if (typeof addressInfo !== 'object' || addressInfo === null) {
+    throw Error('Unexpected server address');
+  }
+  return {
+    port: addressInfo.port,
+    address: addressInfo.address,
+    firstBytes,
+    close: () => server.close(),
+  };
+};
```
Comment (kriskowal) [inline packages/ocapn/test/netlayer-tcp-syrup.test.js:91]: This is odd because it non-deterministically relies on the first packet having a complete message. It no doubt will, but is an unnecessary sensitivity. Consider just capturing the entire session, where the outgoing server sends all of its part of a handshake and then closes immediately. That is, if that is possible to arrange.

File: packages/syrup-frame/test/syrup-frame.test.js [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+        return { done: true };
+      },
+      async throw() {
+        return { done: true };
+      },
+    },
+    opts,
+  );
+  return { array, writer };
+};
```
Comment (kriskowal) [inline packages/syrup-frame/test/syrup-frame.test.js:160]: This seems like a generalizable fixture too. Consider refactoring, or using endo streams more deeply as suggested above.

File: packages/syrup-frame/test/syrup-frame.test.js [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+      return { done: true };
+    },
+    async throw() {
+      return { done: true };
+    },
+  });
+  t.is(writer[Symbol.asyncIterator](), writer);
+});
+
+// -- Adversarial / boundary tests -----------------------------------
```
Comment (kriskowal) [inline packages/syrup-frame/test/syrup-frame.test.js]: ASCII horizontal rule known to be as ill advised as banner. Note in style guide.

File: packages/syrup-frame/CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# Change Log
+
+## 0.1.0 (unreleased)
+
+- Initial release.
+  Provides `makeSyrupReader` and `makeSyrupWriter` for the
+  comma-less length-prefixed framing the OCapN TCP-for-testing
+  protocol has adopted for interoperability testing.
+  Grammar is `<length>:<payload>` with no trailing separator, matching
+  the on-the-wire encoding of a Syrup byte-string record.
```
Comment (kriskowal) [inline packages/syrup-frame/CHANGELOG.md]: Omit. Will get captured by changeset automation.

**Review by kriskowal — CHANGES_REQUESTED** [review]

File: packages/ocapn/src/netlayers/tcp-test-only.js (hunk 3 above)
Comment (kriscendobot) [reply packages/ocapn/src/netlayers/tcp-test-only.js]: Addressed in ff237555c (refactor(ocapn): use @import for @endo/stream Writer/Reader types). Aggregated the gardener-shaped ask (style guide entry, juror checklist or new juror role, discovery automation) into the dispatch report for the steward's follow-up message.

File: packages/ocapn/src/netlayers/tcp-test-only.js (hunk 4 above)
Comment (kriscendobot) [reply packages/ocapn/src/netlayers/tcp-test-only.js]: Addressed in afd81c756 (refactor(ocapn): use slice() to allocate owned frame copy). Replaced the new+set pair with frame.slice(), which yields the same zero-byteOffset owned copy.

File: packages/ocapn/src/netlayers/tcp-test-only.js (hunk 5 above)
Comment (kriscendobot) [reply packages/ocapn/src/netlayers/tcp-test-only.js:325]: Addressed in 109273079 (docs(ocapn): note 'none' framing is debt for the Python test suite). TODO landed on both the TcpTestOnlyFraming typedef and the makeFramedSocketOperations fall-through: default should be 'syrup', 'none' is retained only for the existing Python ocapn-test-suite testing_only_tcp netlayer, and the default flips once that suite either adopts syrup framing or is retired.

File: packages/ocapn/test/netlayer-tcp-syrup.test.js (hunk 6 above)
Comment (kriscendobot) [reply packages/ocapn/test/netlayer-tcp-syrup.test.js]: Addressed in 015e9ab88 (test(ocapn): drop "the JS netlayer" in TCP syrup framing test). Reworded to "the test-only TCP netlayer".

File: packages/ocapn/test/netlayer-tcp-syrup.test.js (hunk 7 above)
Comment (kriscendobot) [reply packages/ocapn/test/netlayer-tcp-syrup.test.js:91]: Addressed in 4b9908321 (test(ocapn): capture full session bytes in syrup framing sniff). The sniffer now accumulates every byte the client writes until the socket closes and resolves a single Promise<Uint8Array>; the assertions on prefix/payload/no-comma are unchanged. The sniffer half-closes its write side on first data so the client tears down its handshake and the close event fires with everything the client managed to send. Verified all three lockdown configs pass locally.

File: packages/syrup-frame/test/syrup-frame.test.js (hunk 8 above)
Comment (kriscendobot) [reply packages/syrup-frame/test/syrup-frame.test.js:160]: Addressed in 27439a6b0 (test(syrup-frame): note makePipe refactor opportunity for makeArrayWriter) with a TODO note rather than the refactor itself. The actual refactor needs an @endo/stream helper (mapReader or similar) that does not yet exist, and the parallel TODO landed in cbor-frame under PR #288 with the same deferral rationale. Will follow up in a separate PR once the helper exists.

File: packages/syrup-frame/test/syrup-frame.test.js (hunk 9 above)
Comment (kriscendobot) [reply packages/syrup-frame/test/syrup-frame.test.js]: Addressed in dcc1a2f49 (test(syrup-frame): drop ASCII section banner). Banner removed along with the two-line sub-heading; per-test comments above each test invocation already name the invariant. Aggregated the gardener-shaped style-guide note into the dispatch report.

File: packages/syrup-frame/CHANGELOG.md (hunk 10 above)
Comment (kriscendobot) [reply packages/syrup-frame/CHANGELOG.md]: Addressed in 8c58e74f8 (chore(syrup-frame): drop the unreleased placeholder CHANGELOG.md). Matches the disposition of the parallel placeholder in cbor-frame, dropped under PR #288. Aggregated the gardener-shaped note (no hand-rolled CHANGELOG on new packages; let changesets emit it) into the dispatch report.

File: .changeset/ocapn-tcp-syrup-framing.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+---
+'@endo/ocapn': minor
+---
+
+- Add an opt-in `framing` option to `makeTcpNetLayer` (`@endo/ocapn/netlayer/tcp-testing`). The default is `'none'`, preserving the current wire format used by the `ocapn/ocapn-test-suite` Python `testing_only_tcp` netlayer (raw syrup-encoded messages with no length prefix). Passing `framing: 'syrup'` wraps each message in the `<length>:<payload>` framing implemented by `@endo/syrup-frame`, which makes the transport robust to TCP chunk boundaries that split a single OCapN message and lets two peers that both opt in share a single length-prefixed primitive with the syrup payload format itself.
```
Comment (kriskowal) [inline .changeset/ocapn-tcp-syrup-framing.md]: Invert the default. The spec is moving toward syrup framing and the test suite has not yet caught up. We’ll enable it explicitly until it’s not needed, then drop it entirely.

File: .changeset/syrup-initial.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+- New package `@endo/syrup-frame`: a sibling of `@endo/netstring` that
+  drops the trailing `,` separator, so each framed payload on the wire
+  is literally a Syrup byte-string record (`<length>:<payload>`).
+- Provides `makeSyrupReader` and `makeSyrupWriter` with the
+  same shape as the netstring equivalents, including the `chunked`
+  zero-copy writer mode.
+- Intended for testing interoperability with the OCapN
+  TCP-for-testing protocol, which has moved to adopt this framing.
+  The framing is among the protocols under consideration in the
+  OCapN pre-standards group at time of writing.
```
Comment (kriskowal) [inline .changeset/syrup-initial.md]: This was settled at the last plenary meeting. Consider citing the relevant minutes.

File: packages/ocapn/src/netlayers/tcp-test-only.js [hunk 13]
```diff
@@ -46,6 +74,102 @@ const makeSocketOperations = (socket, writeLatencyMs) => {
...
+    [Symbol.asyncIterator]() {
+      return sink;
+    },
+  });
+  const syrupWriter = makeSyrupWriter(sink);
+  return {
+    write(bytes) {
+      // The writer's `next` resolves promptly because `sink.next`
+      // resolves on the next microtask; ignore the returned promise.
+      syrupWriter.next(bytes).catch(() => {});
```
Comment (kriskowal) [inline packages/ocapn/src/netlayers/tcp-test-only.js]: That’s a smell. Sync/async impedance mismatch, maybe.

**Review by kriskowal — CHANGES_REQUESTED** [review]

File: .changeset/ocapn-tcp-syrup-framing.md (hunk 11 above)
Comment (kriscendobot) [reply .changeset/ocapn-tcp-syrup-framing.md]: Addressed in bd12c59f6 (fix(ocapn): invert tcp-test-only default to syrup framing). The default is now `framing: 'syrup'`; the Python `ocapn-test-suite` driver (`packages/ocapn/test/python-test-suite/index.js`) opts into `framing: 'none'` explicitly. The changeset and the `TcpTestOnlyFraming` typedef are reworded to lead with the new default and to describe `'none'` as the temporary opt-in retained only for that suite's sake.

File: .changeset/syrup-initial.md (hunk 12 above)
Comment (kriscendobot) [reply .changeset/syrup-initial.md]: Addressed in e55ae0187 (docs(syrup-frame): cite 2025-12-09 OCapN plenary on TCP-for-testing framing). Cited the 2025-12-09 plenary minutes (https://github.com/ocapn/ocapn/blob/main/meeting-minutes/2025-12-09.md) and the parallel discussion on ocapn/ocapn#104 (your comment of that date capturing the converged shape: "a Syrup stream that is limited to byte arrays which themselves subcontain arbitrary Syrup messages"). The changeset prose now records that the plenary settled on the length-prefixed Syrup byte string framing this package implements. Cross-cited the same plenary on the inverted-default changeset for ask #1, so a reader of either changelog has a pointer to the same provenance.

File: packages/ocapn/src/netlayers/tcp-test-only.js (hunk 13 above)
Comment (kriscendobot) [reply packages/ocapn/src/netlayers/tcp-test-only.js]: Addressed in 8f9d7b438 (refactor(ocapn): drop async indirection in syrup-framing socket writer). The smell was real: `SocketOperations.write` is synchronous (returns `void`), but the implementation chained through a `@endo/syrup-frame` async `Writer` over a microtask-resolving sink, then discarded the returned promise with `.catch(() => {})`. The async indirection bought nothing (the sink's `next` synchronously called `socketOps.write` before yielding the promise) and silently swallowed any error the framing step or sink might have raised. The refactor builds the `<length>:` prefix with `TextEncoder`, concatenates one buffer of prefix + payload, and forwards a single `socketOps.write(frame)` call. Socket-level errors continue to surface through the `socket.on('error')` handler in `setupSocketHandlers`. Tests in `netlayer-tcp-syrup.test.js` (framing-bytes-on-wire, round-trip, unknown-framing-rejection) pass unchanged under all three lockdown configs; full `packages/ocapn` suite (260 tests) green.

**Comment by kriskowal** [comment]
Ferried the six follow-ups onto `endojs/endo#3256` as a fast-forward append on top of `b1cf4aff`. Upstream tip is now `e691e86d8`.

Six new commits in order:

1. `639a29205` test(syrup-frame): note makePipe refactor opportunity for makeArrayWriter
2. `870c3befe` test(syrup-frame): drop ASCII section banner
3. `6cd3f6b32` chore(syrup-frame): drop the unreleased placeholder CHANGELOG.md
4. `0ce0d61f5` fix(ocapn): invert tcp-test-only default to syrup framing
5. `885e377c4` docs(syrup-frame): cite 2025-12-09 OCapN plenary on TCP-for-testing framing
6. `e691e86d8` refactor(ocapn): drop async indirection in syrup-framing socket writer

No force-push; the prior three commits on the upstream tip are unchanged.

**Comment by kriskowal** [comment]
Please reset and restage these changes into grouped commits per standing-instructions.

**Comment by kriskowal** [comment]
Recomputed onto upstream master and force-pushed with lease to endojs/endo#3256 `feat/syrups-package`. New head: `f5182df17`. Four commits in order:

- `3d4d265fe` feat(syrup-frame): add @endo/syrup-frame package
- `eaa96182c` feat(ocapn): add opt-in syrup framing to TCP-testing netlayer
- `f9ea9accd` chore: Update yarn.lock
- `f5182df17` chore: regenerate composite tsconfig files

Replaces the prior 9-commit append shape (`e691e86d8`, the lease tip). Push used `--force-with-lease`. kumavis's APPROVED review persists (anchored at `ed80869d4`).

Note on naming: the upstream PR title carries plural 'syrups' (`opt-in syrups framing`) and the branch is `feat/syrups-package`, but the code paths and identifiers throughout are singular: package `@endo/syrup-frame`, imports from `@endo/syrup-frame/reader.js`, `makeSyrupReader`/`makeSyrupWriter`/`makeSyrupDeframer`, framing option value `'syrup'`, test file `netlayer-tcp-syrup.test.js`. The title/branch are the outliers; the code is consistent.

**Comment by kriskowal** [comment]
Recomputed onto fresher `endojs/endo` master tip `ec3dcbc0` (advanced 11 commits since the prior ferry); source head `2627e81a` unchanged. Four-commit shape preserved, all attributed to `Kris Kowal <kriskowal@kriskowal.com>`. Upstream PR endojs/endo#3256 now at `da5a9b48edbc6a0609a3fd4086676e797008dacb`. kumavis APPROVED persists.

**Comment by kriskowal** [comment]
Merged upstream https://github.com/endojs/endo/pull/3256. Please dispatch a weaver to merge actual/master into bots/llm.
